### PR TITLE
Develop facebook

### PR DESCRIPTION
**Pull request by Arnvig**

### CODE REVIEW

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+{
+    public FBPicture data { get; set; }
+}
+    
+public class FBPicture
+{
+    public string url { get; set; }
+    public int height { get; set; }
+    public int width { get; set; }
+    public bool is_silhouette { get; set; }
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:63]: Fix naming to fit the C# convention

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+}
+    
+public class FBidRespons
+{
+    public string id { get; set; }
+}
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:69]: Doesn't follow C# naming conventions

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections.Generic;
+
+namespace Dikubot.DataLayer.Database.Global.Facebook;
+
+
+public class FacebookPage : MainModel
+{
+    
+    public FBPages Item { get; set; }
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:9]: FacebookPage only contains a FBPages object. Why not make it 1 object instead of 2 different?

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+public class FBrespons
+{
+    public List<FBPages> data { get; set; }
+}
+    
+public class FBPages
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:17]: "FBPages" is plural, but this is a singular page, isn't it?

File: Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor [hunk 5]
```diff
@@ -0,0 +1,144 @@
+﻿@page "/FacebookAddPages"
+@using System.Net.Http.Json
+@using System.Threading
+@using Dikubot.DataLayer.Database.Global.Facebook
+@inject FacebookPageService _FacebookPageService
+@inject NavigationManager _navigationManager
+
+@attribute [AllowAnonymous]
+<h3>fb</h3>
```
Comment (Lukires) [inline Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor:9]: Did you forget to remove the header?

File: Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    private string GetFBPicture(string id)
+    {
+        if (!string.IsNullOrWhiteSpace(AuthCode))
+        {
+            var link = $"https://graph.facebook.com/v16.0/{id}/picture?redirect=0&access_token={AuthCode}";
+            return HttpClient.GetFromJsonAsync<FBPictureRespons>(link).Result.data.url;
+
+    //return ;
+        }
+        return "";
```
Comment (Lukires) [inline Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor:65]: Most of this should be moved into a service. Consider what I have previously mentioned when it comes to implementing services with dependency injection.

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookPageService.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Dikubot.DataLayer.Database.Global.Facebook;
+
+public class FacebookPageService : GlobalMongoService<FacebookPage>
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookPageService.cs:3]: See the naming conventions we have used for database services. Consider changing it to FacebookPageDatabaseService

File: Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor (hunk 6 above)
Comment (Lukires) [reply Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor:65]: Same can be said for a lot of the code under this.

File: Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+           var respons3 = HttpClient.GetFromJsonAsync<FBAccessToken>(link3).Result;
+
+            var link4 = $"https://graph.facebook.com/me?fields=id&access_token={respons3.access_token}";
+            var app_scoped_user_id = HttpClient.GetFromJsonAsync<FBidRespons>(link4).Result.id;
+            
+            var link5 = $"https://graph.facebook.com/{graph_api_version}/{app_scoped_user_id}/" +
+                        $"accounts?access_token={respons3.access_token}";
+
+            var test5 = HttpClient.GetFromJsonAsync<FBrespons>(link5).Result.data;
+
```
Comment (Lukires) [inline Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor:122]: The variable names don't follow the C# naming convention, also they are not very helpful. Try to be descriptive in the names. What is link5, which uses response3? It is hard to read.

**Review by Lukires — CHANGES_REQUESTED** [review]
See comments.

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs (hunk 1 above)
Comment (Arnvig) [reply Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:63]: Sinces it's from a json - i do not know if i just can fix it

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs (hunk 2 above)
Comment (Arnvig) [reply Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:69]: Sinces it's from a json - i do not know if i just can fix it

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs (hunk 3 above)
Comment (Arnvig) [reply Dikubot/DataLayer/Database/Global/Facebook/FacebookModels.cs:9]: Since the FacebookPage have ID as GUID, an FBPage have an ID as string/numbers

File: Dikubot/DataLayer/Database/Global/Facebook/FacebookPageDatabaseService.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        var accessToken = this.Get(m => m.Item.Id == id)?.Item?.AccessToken;
+        
+        if (!string.IsNullOrWhiteSpace(accessToken))
+        {
+            var link = $"https://graph.facebook.com/v16.0/{id}/picture?redirect=0&access_token={accessToken}";
+            return httpClient.GetFromJsonAsync<FBPictureRespons>(link).Result.Data.Url;
+            
+        }
+        return "";
+    }
```
Comment (Lukires) [inline Dikubot/DataLayer/Database/Global/Facebook/FacebookPageDatabaseService.cs:47]: This is a database service. This logic is not related to the database, and should therefore not be here. Make a FacebookService.cs in webapp/data. Look at some other of the services in that folder for inspiration.

File: Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                            $"client_id={appId}&" +
+                            $"client_secret={appSecret}&" +
+                            $"fb_exchange_token={shortLivedAccessToken.AccessToken}";
+                var longLivedUserAccessToken = HttpClient.GetFromJsonAsync<FBAccessToken>(longLivedAccessTokenLink).Result;
+                
+                    
+                // Link to get the unique app User id
+                var userIdLink = $"https://graph.facebook.com/me?fields=id&access_token={longLivedUserAccessToken.AccessToken}";
+                var appScopedUserId = HttpClient.GetFromJsonAsync<FBidRespons>(userIdLink).Result.Id;
+
```
Comment (Lukires) [inline Dikubot/Webapp/Shared/Facebook/FacebookAddPages.razor:109]: It seems like a lot of this could be added to the aforementioned FacebookService.cs

**Review by Lukires — CHANGES_REQUESTED** [review]
Looking very good! Only a few small requests.

File: Dikubot/Webapp/Data/FacebookService.cs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            
+        }
+        return "";
+    }
+
+    public void ConnectFacebookAction(string redirectUri)
+    {
+        string link =
+            $"https://www.facebook.com/{graphApiVersion}/dialog/oauth?" +
+            "client_id=" + Environment.GetEnvironmentVariable("FACEBOOK_CLIENT_ID") +
```
Comment (Lukires) [inline Dikubot/Webapp/Data/FacebookService.cs:64]: Move Environment.GetEnvironmentVariable("FACEBOOK_CLIENT_ID")  out of here. It is a constant.

File: Dikubot/Webapp/Data/FacebookService.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        var accessToken = _facebookPageService.Get(m => m.PageId == id).AccessToken;
+        
+        var requestUri = $"https://graph.facebook.com/{graphApiVersion}/{id}/events?access_token={accessToken}";
+        var result = _httpClient.GetFromJsonAsync<FBRespons<FBEvent>>(requestUri).Result.Data;
+        FBEventDic.Add(id, result);
+        return result;
+    }
+    
+    public string GetFBPicture(string id)
```
Comment (Lukires) [inline Dikubot/Webapp/Data/FacebookService.cs:46]: Since id is referring to a page id, perhaps have it accept the FacebookPage object instead, such that mistakes are less likely, and then get  the id from the FacebookPage object. This can also be said for the method GetFBEvents(string id)

File: Dikubot/Webapp/Data/FacebookService.cs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+                return _httpClient.GetFromJsonAsync<FBRespons<FBPage>>(longLivedPageAccessTokenLink).Result.Data;
+                
+                
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
+                throw;
+            }
```
Comment (Lukires) [inline Dikubot/Webapp/Data/FacebookService.cs:119]: This is bad error handling. It would be better to handle the errors and have good and understandable errors passed to the users. I will let it slide in this case, but keep it in mind.

**Review by Lukires — APPROVED** [review]
